### DISCUSSION TITLE
Next.js 14: How to pass additional params that are not used for routing with generateStaticParams (SSG strategy)

### DISCUSSION THREAD

**Discussion by danilo-timacum** [vercel/next.js#58549] [Help]
### Summary

I've configured my blog to use slugs for routing post routes, but static data retrieval relies on unique IDs. I need a method to pass these IDs in a way that doesn't affect routing but allows data fetching for each page and with slug as page route during the build process.

Is this possible in NextJS 14?

This is what I am trying to do:
```
import { fetchPage } from '@/lib/requests'
import { notFound } from 'next/navigation'

// Return a list of `params` to populate the [slug] dynamic segment
export async function generateStaticParams() {
    const page = await fetchPage('/posts?populate=deep')

    return page.data.map((post) => ({
        slug: post.attributes.Seo.Url,
        id: post.id, // this will not work
    }))
}


export default async function Post({params}) {
    const { id } = params

    const page = await fetchPage(`/posts/${id}?populate=deep`)

    if (!page) {
        return notFound()
    }

    return <h1>{id}</h1>
}
```

**Comment by j0hnm4r5** [comment]
Another use-case: if I have to do an API call to the CMS to get all of my posts (and extract their slugs to pass to `generateStaticParams`), why do I have to do essentially the same API call from the page when it gets generated? Why not just pass the entire schema over params from the start?

**Comment by j0hnm4r5** [comment]
@danilo-timacum I'm by no means an expert, but as it stands now I don't think it's possible. You'll probably need to write an API call that maps slug to post ID (or remove either post ID or slug from the schema altogether, as you probably don't need two different ways to uniquely identify a post).

**Comment by levipadre** [comment]
I have the same issue, but on top of that, it's a multilingual site and the URL is not simply /en or /fr, but domain.com, domain.uk, domain.fr etc.
I really want to use `generateStaticParams`, but I can't. I have site and page IDs, but I can't match with the pages because the slug is just part of the URL.
I stuck with `headers()` to get the host information atm, but it would be great to have a solution for such cases.
@timneutkens, can't that be a feature in Nextjs? Sorry I marked you directly, but it seemed an important thing.

**Comment by marceloverdijk** [comment]
I have the same question...

Besides the `id` I want to pass additional attributes to the `Page` `params` and mainly the `generateMetadata` `params`.

Now I have to query the name in the `generateMetadata` again.


``` 
import type { Metadata } from 'next'
import prisma from '@/lib/db'
import { Container } from 'react-bootstrap'

export async function generateStaticParams() {
  const drivers = await prisma.driver.findMany({
    select: {
      id: true,
    },
  })
  return drivers.map((driver) => ({
    id: driver.id,
  }))
}

export async function generateMetadata({ params }: { params: { id: string, name: string } }): Promise<Metadata> {
  const driver = await prisma.driver.findUniqueOrThrow({
    select: {
      name: true,
    },
    where: {
      id: params.id,
    },
  })
  return {
    title: driver.name,
  }
}

export default async function Page({ params }: { params: { id: string, name: string } }) {
  console.log(params)
  const driver = await prisma.driver.findUniqueOrThrow({
    where: {
      id: params.id,
    },
  })
  return (
    <Container>
      <h1>{driver.name}</h1>
      <div>
        <pre className="overflow-auto" style={{ height: '50rem' }}>
          {JSON.stringify(driver, null, 2)}
        </pre>
      </div>
    </Container>
  )
}


```

**Comment by grazianodev** [comment]
Am interested in this too, would like to achieve exactly what OP said.

**Comment by petegarnett** [comment]
Also really interested in this.I'd like to build my structure off slugs and then I need to get my data from contentful with the ID.

**Comment by Pavan-Sai-Grandhi** [comment]
Is there any solution for this? I am also facing the same issue as @marceloverdijk. Calling the same API multiple times feels redundant.

**Comment by freemvmt** [comment]
Another great example of when this would be super useful is if you're using Airtable as a backend (as is common for MVPs built off existing data etc), and you want your URL to show your primary key*, but you want to also pass down the Airtable record ID proper (`recxxxxxxx`) for querying the data in each page without making a silly number of requests, which they then rate limit.

* usually composed of some other fields glued together to make a unique ID of some kind

Not sure what I'm gonna do for a workaround yet - maybe concat them in the slug and pull the record off, but that's not particularly aesthetic 🤔 

NB. I'm using Next15 with the app router
